### PR TITLE
Refactor `cache_execute()` as a transform

### PR DESCRIPTION
**Pull request by Mandrenkov**
**Context:**

Presently, the [`cache_execute()`](https://docs.pennylane.ai/en/stable/code/api/pennylane.workflow.cache_execute.html?highlight=cache_execute#pennylane.workflow.cache_execute) function is difficult to understand because it assumes too many responsibilities. This PR offers an alternative implementation, `_cache_transform()`, which leverages the [`@transform`](https://docs.pennylane.ai/en/stable/code/api/pennylane.transforms.core.transform.html) infrastructure to provide the same functionality but in a shorter, cleaner, and more concise way.

If this PR is approved, we should consider deprecating `cache_execute()` in the next release.

**Description of the Change:**

* Added the following functions:
    * `_cache_transform()` which re-implements `cache_execute()` using `@transform`.
    * `_apply_cache_transform()` which conveniently wraps `_cache_transform()`.
* Updated the existing `cache_execute()` tests to spy on `_cache_transform()` instead.
* Added a new set of tests to explicitly test the behaviour of `_cache_transform()`.

**Benefits:**

* A significant developer productivity boost for those debugging `cache_execute()`.

**Possible Drawbacks:**

* Some additional work will be required to deprecate and eventually remove `cache_execute()`.

**Related GitHub Issues:**

None.

### CODE REVIEW

**Comment by Mandrenkov** [comment]
@albi3ro I've tagged you since you are the primary code owner but I think we should get someone else involved too.

File: pennylane/workflow/execution.py
Comment (albi3ro) [inline pennylane/workflow/execution.py:266]: I think at this point we can just directly use `_cache_transform` inside `inner_execute`.

```
new_batch, postprocessing = _cache_transform(tapes)
device_results = device_execution(new_batch)
postprocessing(results)
```


I'd like to move toward a place where we have an "Inner transform program":

```
program = TransformProgram()
if expand_fn:
    program.add_transform(expand_fn_transform(expand_fn))
if numpy_only:
    program.add_transform(convert_to_numpy_parameters)
program.add_transform(_cache_transform)

new_batch, postprocessing = program(tapes)
device_results = device_execution(new_batch)
postprocessing(device_results)
```

We would just need to convert `convert_to_numpy_parameters` to a transform and duplicate the `expand_fn_transform` code from `workflow/construct_batch.py`. So eventually we'll get there.

Comment (Mandrenkov) [reply pennylane/workflow/execution.py:266]: @albi3ro I think that suggestion makes a lot of sense! That was actually my original approach but I ended up going with `_apply_cache_transform()` to reuse the same code when calculating `"device"` gradients in `execute()` (see `execution.py` line ~800).

As such, do you think it makes sense to adopt the `TransformProgram` style in `_make_inner_execute()` but use `_apply_cache_transform()` in the device gradient `else` branch? Or should we use it in both places? 

We could also potentially keep `_apply_cache_transform()` for now and switch to using `TransformProgram` when we get around to rewriting `convert_to_numpy_parameters()` as a transform.

**Comment by Mandrenkov** [comment]
Thanks for the comment, @albi3ro!

File: pennylane/workflow/execution.py
Comment (albi3ro) [reply pennylane/workflow/execution.py:266]: So the `"device"` gradients logic is now in low-maintainence mode, and at some point I want to just relegate all that logic into `tensorflow_autograph.py` and let it deal with itself. So the philosophy with that code is "keep it working with the minimal amount of thought and effort".

But for `_make_inner_execute`, I think it is worth moving it in the direction we want it to be, even if that is slowly.

But I'm good with both approaches.  This file will 🤞 improve someday.

**Comment by Mandrenkov** [comment]
Thanks for the updated review, @albi3ro! 🥳

**Review by albi3ro — COMMENTED** [review]
OK I think I see what the failures are.  `cache=None` should be a valid kwarg to `_cache_transform`, in which case we just return `(tape,) , null_postprocessing` early.

**Comment by Mandrenkov** [comment]
> OK I think I see what the failures are. `cache=None` should be a valid kwarg to `_cache_transform`, in which case we just return `(tape,) , null_postprocessing` early.

Ah, sorry @albi3ro, I missed that the tests failed!

In my opinion, we shouldn't support passing `None` to `_cache_transform` because it can hide bugs and encourage bad programming practices: if you don't have a cache, you shouldn't be calling `_cache_transform` in the first place.

File: pennylane/workflow/execution.py
Comment (albi3ro) [inline pennylane/workflow/execution.py:298]:
```suggestion
        transformed_tapes, transform_post_processing = transform_program(tapes)
        if len(transformed_tapes) == 0:
            return transform_post_processing(tuple())
        return transform_post_processing(device_execution(transformed_tapes))
```

This should fix the failures in `tests/interfaces/default_qubit_2_integration/test_execute_default_qubit_2.py` .

File: tests/interfaces/test_jax.py
Comment (albi3ro) [inline tests/interfaces/test_jax.py:198]:
```suggestion
        spy = mocker.spy(qml.workflow.execution._cache_transform, "_transform")
```

The mocking should work with this.

File: tests/interfaces/test_jax.py
Comment (albi3ro) [inline tests/interfaces/test_jax.py:225]:
```suggestion
        spy = mocker.spy(qml.workflow.execution._cache_transform, "_transform")
```

File: tests/interfaces/test_jax.py
Comment (albi3ro) [inline tests/interfaces/test_jax.py:251]:
```suggestion
        spy = mocker.spy(qml.workflow.execution._cache_transform, "_transform")
```

File: tests/interfaces/test_jax_jit.py
Comment (albi3ro) [inline tests/interfaces/test_jax_jit.py:187]:
```suggestion
        spy = mocker.spy(qml.workflow.execution._cache_transform, "_transform")
```

File: tests/interfaces/test_jax_jit.py
Comment (albi3ro) [inline tests/interfaces/test_jax_jit.py:215]:
```suggestion
        spy = mocker.spy(qml.workflow.execution._cache_transform, "_transform")
```

File: tests/interfaces/test_jax_jit.py
Comment (albi3ro) [inline tests/interfaces/test_jax_jit.py:242]:
```suggestion
        spy = mocker.spy(qml.workflow.execution._cache_transform, "_transform")
```

File: tests/interfaces/test_tensorflow.py
Comment (albi3ro) [inline tests/interfaces/test_tensorflow.py:140]:
```suggestion
        spy = mocker.spy(qml.workflow.execution._cache_transform, "_transform")
```

File: tests/interfaces/test_tensorflow.py
Comment (albi3ro) [inline tests/interfaces/test_tensorflow.py:162]:
```suggestion
        spy = mocker.spy(qml.workflow.execution._cache_transform, "_transform")
```

File: tests/interfaces/test_torch.py
Comment (albi3ro) [inline tests/interfaces/test_torch.py:174]:
```suggestion
        spy = mocker.spy(qml.workflow.execution._cache_transform, "_transform")
```

File: tests/interfaces/test_torch.py
Comment (albi3ro) [inline tests/interfaces/test_torch.py:200]:
```suggestion
        spy = mocker.spy(qml.workflow.execution._cache_transform, "_transform")
```

**Comment by albi3ro** [comment]
Left suggestions that fix the failures due to extra batches and mocking.

The remaining failures are due to a different number of logs that occur. We could either reduce the number of expected logs, or add logging to the transform.  Both options would be equally fine to me.

File: pennylane/workflow/execution.py
Comment (trbromley) [inline pennylane/workflow/execution.py:97]: Is this only typically expected when a user is directly using `qml.execute()`?

Comment (Mandrenkov) [reply pennylane/workflow/execution.py:97]: Nope, @trbromley!

You can easily trigger it using a `QNode` too:
```python

import pennylane as qml

dev = qml.device("default.qubit", shots=1)

@qml.qnode(dev, cache={})
def circuit():
    return qml.expval(qml.PauliZ(0))

print(circuit())
print(circuit())
```
Running the script above gives:
```
1.0
/Users/mikhail/sc/39484/pennylane/pennylane/workflow/execution.py:462: UserWarning: Cached execution with finite shots detected!
Note that samples as well as all noisy quantities computed via sampling will be identical across executions. This situation arises where tapes are executed with identical operations, measurements, and parameters.
To avoid this behaviour, provide 'cache=False' to the QNode or execution function.
  warnings.warn(_CACHED_EXECUTION_WITH_FINITE_SHOTS_WARNINGS, UserWarning)
1.0
```

**Comment by Mandrenkov** [comment]
Thanks for all the suggestions, @albi3ro!

I've learned my lesson now which is to wait until CI finishes before re-requesting a review. 😅

**Comment by albi3ro** [comment]
> Thanks for all the suggestions, @albi3ro!
> 
> I've learned my lesson now which is to wait until CI finishes before re-requesting a review. 😅

I definitely do that all the time myself 😬

**Review by albi3ro — APPROVED** [review]
So the code cov failures are strictly due the lack of unit tests for `cache_execute`.  Given we are more or less marking that function for removal, I'm fine with leaving those lines uncovered.  Though this would potentially be another piece of evidence for removing `cache_execute` immediately.

But the new transform is so much easier to follow than `cache_execute` 🎉 Thank you so much for handling this improvement :)

**Comment by Mandrenkov** [comment]
Thanks, @albi3ro. Any advice on how to deal with the decline in CodeCov coverage? The issue mostly stems from the fact that `cache_execute()` is not called anymore.

**Comment by Mandrenkov** [comment]
Thanks for the many reviews and advice, @albi3ro!

**Comment by Mandrenkov** [comment]
After offline discussion, it was decided to remove `cache_execute()` altogether.

File: pennylane/workflow/jacobian_products.py
Comment (trbromley) [inline pennylane/workflow/jacobian_products.py:226]:
```suggestion
            by the cache.
```
To minimize advertising this function in the docs?

Comment (Mandrenkov) [reply pennylane/workflow/jacobian_products.py:226]: Good idea. Thanks! 🎉

**Comment by PietropaoloFrisoni** [comment]
Thanks, @Mandrenkov. This looks fantastic! 
I would only suggest merging from master after 8 days to ensure everything is OK (except for the issue with lightning, which is most probably independent of this PR) :+1:

**Review by albi3ro — APPROVED** [review]
Reapproving after the removal after removal of `cache_execute`.

**Review by PietropaoloFrisoni — APPROVED** [review]
Amazing! Thanks again @Mandrenkov

**Pull request merged**
